device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java

```diff
@@ -487,39 +493,45 @@ public void close(IotHubConnectionStatusChangeReason reason, Throwable cause)
             throw new IllegalArgumentException("reason cannot be null");
         }
 
-        this.cancelPendingPackets();
+        this.isClosing = true;
 
-        this.invokeCallbacks();
-
-        if (this.taskScheduler != null)
+        // Wait until no reconnection logic is taking place
+        synchronized (this.reconnectionLock)
```

Comment:
Having locks within a lock makes me worried about deadlocks. Thoughts?

Reply:
Are you referring to the ```sendThreadLock``` and ```receiveThreadLock```? Neither of those locks are acquired or modified in the reconnection thread. There is no chance that the reconnection thread holds onto the sendThreadLock or receiveThreadLock while the ```close()``` thread holds the reconnectionLock so we're safe from deadlock